services/QuillLMS/app/services/classlink_integration/enrollment_persister.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module ClasslinkIntegration
+  class EnrollmentPersister < ApplicationService
+    attr_reader :incoming_enrollment, :existing_provider_users, :existing_provider_classrooms
+
+    COTEACHER = ClassroomsTeacher::ROLE_TYPES[:coteacher]
+    OWNER = ClassroomsTeacher::ROLE_TYPES[:owner]
+
+    def initialize(incoming_enrollment:, existing_provider_users:, existing_provider_classrooms:)
+      @incoming_enrollment = incoming_enrollment
+      @existing_provider_users = existing_provider_users
+      @existing_provider_classrooms = existing_provider_classrooms
+    end
+
+    def quill_user
+      @quill_user ||= existing_provider_users.find do |eu|
+        eu.external_id == incoming_enrollment[:user_sourced_id]
+      end&.user
+    end
+
+    def quill_classroom
+      @quill_classroom ||= existing_provider_classrooms.find do |ec|
+        ec.external_id == incoming_enrollment[:classroom_sourced_id]
+      end&.classroom
+    end
+
+    def run
+      return unless quill_classroom && quill_user
+
+      case quill_user.role
+      when ::User::TEACHER
+        create_classroom_teacher(quill_user, quill_classroom)
+      when ::User::STUDENT
+        # TODO: (probably)
+        # bifurcate the EnrollmentPersister into two classes, one for teachers and one for students
+        # We cannot use StudentClassroomAssociator unless we process all teachers before all students
+        # ::StudentClassroomAssociator.run(quill_user, quill_classroom)
+        ActiveRecord::Base.transaction do
+          StudentsClassrooms.unscoped.find_or_create_by!(student: quill_user, classroom: quill_classroom)
+
+          ClasslinkClassroomUser.find_or_create_by!(
+            classroom_external_id: incoming_enrollment[:classroom_sourced_id],
+            user_external_id: incoming_enrollment[:user_sourced_id]
+          )
+        end
```

Comment:
Is it worth putting this in a method `create_classroom_student` or something to match the pattern of `create_classroom_teacher`?

Reply:
Yes, except that I'm replacing this code block with an Associator - see comment